src/main/java/com/hrznstudio/titanium/material/ResourceMaterial.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * This file is part of Titanium
+ * Copyright (C) 2019, Horizon Studio <contact@hrznstudio.com>.
+ *
+ * This code is licensed under GNU Lesser General Public License v3.0, the full license text can be found in LICENSE.txt
+ */
+
+package com.hrznstudio.titanium.material;
+
+import com.hrznstudio.titanium.api.material.IResourceType;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+
+public class ResourceMaterial {
+
+    private String type;
+    private HashMap<String, IResourceType> generatorTypes;
+    private HashMap<String, ForgeRegistryEntry> generatorOverrides;
+    private HashMap<String, ForgeRegistryEntry> generated;
+    private int color;
+
+    ResourceMaterial(String type) {
+        this.type = type;
+        this.generatorTypes = new HashMap<>();
+        this.generatorOverrides = new HashMap<>();
+        this.generated = new HashMap<>();
+    }
+
+    public String getMaterialType() {
+        return type;
+    }
+
+    public ResourceMaterial add(IResourceType type) {
+        generatorTypes.computeIfAbsent(type.getTag(), s -> type);
+        return this;
+    }
+
+    public ResourceMaterial addAll(IResourceType... type) {
+        for (IResourceType iResourceType : type) {
+            add(iResourceType);
+        }
+        return this;
+    }
+
+    public ResourceMaterial addOverride(IResourceType type, ForgeRegistryEntry entry) {
```

Comment:
Generally if you're doing methods like a builder "With" to show that it's adding pieces to a builder.